### PR TITLE
added stage position legend

### PR DESCRIPTION
**Pull request by nandishjpatel**
Ticket: https://delmic.atlassian.net/browse/SDEV-1377

### CODE REVIEW

**Review by pieleric — CHANGES_REQUESTED** [review]
For now it adds a big extra info to an already crowded legend... for every viewport of the GUI. Even for the viewports which are not connected a real stage (such as in the analysis tab). I think for now it should be limited to the FASTEM GUI.

**Comment by nandishjpatel** [comment]
> For now it adds a big extra info to an already crowded legend... for every viewport of the GUI. Even for the viewports which are not connected a real stage (such as in the analysis tab). I think for now it should be limited to the FASTEM GUI.

Should I add the stage position subscriber to FAST-EM classes [here](https://github.com/delmic/odemis/blob/aaab22a46d8a061266098765b7cbd1bc6e226172/src/odemis/gui/comp/viewport.py#L832-L884) instead of MicroscopeViewport?

File: src/odemis/gui/comp/legend.py [hunk 1]
```diff
@@ -125,6 +125,12 @@ def __init__(self, parent, wid=-1, pos=(0, 0), size=wx.DefaultSize,
         self.zPos_text.SetToolTip("Z Position")
         self.zPos_text.Hide()
 
+        # Stage position
+        self.stage_pos_text = wx.TextCtrl(self, style=wx.NO_BORDER | wx.CB_READONLY)
+        self.stage_pos_text.SetBackgroundColour(parent.GetBackgroundColour())
+        self.stage_pos_text.SetForegroundColour(parent.GetForegroundColour())
+        self.stage_pos_text.SetToolTip("Current Stage Position")
+
```
[the remark this answers is not in the document]
Comment (nandishjpatel) [reply src/odemis/gui/comp/legend.py]: Yes, I didn't know the consequences

**Review by pieleric — CHANGES_REQUESTED** [review]

File: src/odemis/gui/comp/viewport.py [hunk 2]
```diff
@@ -324,6 +324,11 @@ def setView(self, view, tab_data):
         if model.hasVA(tab_data, "zPos"):
             tab_data.zPos.subscribe(self._on_zPos_change, init=True)
 
+        # Listen to the stage position, if the view has the stage and the stage position TextCtrl
+        # has been set to be shown (by default it is hidden)
+        if hasattr(view, "stage_pos") and self.bottom_legend.stage_pos_text.IsShown():
+            view.stage_pos.subscribe(self._on_stage_pos_change, init=True)
+
```
Comment (pieleric) [inline src/odemis/gui/comp/viewport.py:331]: That feels reversed organization. Now the child class needs to make sure the stage_pos_text is shown, before calling the parent setView. Moreover the idea of `set_stage_pos_label()` is that it takes care of hidding/showing the internal widgets.
I'd suggest to entirely remove the code from this class. Then, in FastEM*Viewport classes, override `.setView()` to add:
```python
if hasattr(view, "stage_pos"):
        view.stage_pos.subscribe(self._on_stage_pos_change, init=True)
```
(the `if` might not be necessary, as normally there is always a stage on the FASTEM)

**Review by pieleric — CHANGES_REQUESTED** [review]

File: src/odemis/gui/comp/viewport.py (hunk 2 above)
Comment (nandishjpatel) [reply src/odemis/gui/comp/viewport.py:331]: Ahh, I misunderstood your previous comment

**Review by pieleric — APPROVED** [review]

**Pull request merged**